src/js_parser.zig

```diff
@@ -17355,9 +17356,151 @@ fn NewParser_(
+
+                            instanceMembers.append(p.s(S.SExpr{ .value = Expr.assign(target, initialValue, p.allocator) }, loc)) catch unreachable;
+                        }
+
+                        // handle methods
+                        if (prop.flags.contains(.is_method)) {
+                            if (prop.value.?.data == .e_function and prop.key.?.data == .e_string and
+                                prop.key.?.data.e_string.eqlComptime("constructor"))
+                            {
+                                ctor = prop.value.?.data.e_function;
+                            }
+                            updatedClassProps.append(prop) catch unreachable;
+                        }
+                    }
+
+                    class.properties = updatedClassProps.toOwnedSlice();
+
+                    if (instanceMembers.items.len > 0) {
+                        // create constructor if needed
+                        if (ctor == null) {
+                            ctor = &(E.Function{
+                                .func = G.Fn{
+                                    .name = null,
+                                    .open_parens_loc = logger.Loc.Empty,
+                                    .args = &([_]Arg{}),
+
+                                    // might need super() before these statements
+                                    .body = .{ .loc = stmt.loc, .stmts = instanceMembers.toOwnedSlice() },
+                                    .flags = Flags.Function.init(.{}),
+                                },
+                            });
+
+                            var properties = ListManaged(G.Property).fromOwnedSlice(p.allocator, class.properties);
+
+                            // move ctor to first index
+                            properties.insert(0, G.Property{
+                                .flags = Flags.Property.init(.{ .is_method = true }),
+                                .key = p.e(E.String{ .data = "constructor" }, stmt.loc),
+                                .value = p.e(ctor.?, stmt.loc),
+                            }) catch unreachable;
+
+                            class.properties = properties.toOwnedSlice();
+                        } else {
+                            // constructor already exists, append initializers
+                            var ctorStmts = ListManaged(Stmt).fromOwnedSlice(p.allocator, ctor.?.func.body.stmts);
+                            ctorStmts.appendSlice(instanceMembers.toOwnedSlice()) catch unreachable;
+                            ctor.?.func.body.stmts = ctorStmts.toOwnedSlice();
+                        }
+                    }
+
+                    // append class
+                    if (class.ts_decorators.len > 0) {} else {
+                        stmts.append(p.s(S.Class{ .class = class }, class.class_name.?.loc)) catch unreachable;
+                    }
+
+                    stmts.appendSlice(instanceDecorators.toOwnedSlice()) catch unreachable;
```

Comment:
lets avoid using `toOwnedSlice` and instead use `.items` directly because `toOwnedSlice` causes the memory to resize, which shows up in profiling sometimes

Reply:
👍